cmd/duffle/relocate.go

```diff
@@ -0,0 +1,249 @@
+	bun, err := r.setup()
+	if err != nil {
+		return err
+	}
+
+	if err := r.relocate(bun); err != nil {
+		return err
+	}
+
+	if _, err := r.writeBundle(bun); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (r *relocateCmd) relocate(bun *bundle.Bundle) error {
+	// mutate the input bundle to become the output bundle
+	if !r.outputBundleIsFile {
+		bun.Name = r.outputBundle
+	}
+
+	ks := keys(bun.Images)
+	for _, k := range ks {
+		im := bun.Images[k]
+		if isOCI(im.ImageType) || isDocker(im.ImageType) {
+			// map the image name
+			n, err := image.NewName(im.Image)
+			if err != nil {
+				return err
+			}
+			rn := r.mapping(r.repoPrefix, n)
+
+			// tag/push the image to its new repository
+			dig, err := r.registryClient.Copy(n, rn)
+			if err != nil {
+				return err
+			}
+			if dig.String() != im.Digest {
+				// should not happen
+				return fmt.Errorf("digest of image %s not preserved: old digest %s; new digest %s", im.Image, im.Digest, dig.String())
+			}
+
+			// update the imagemap
+			im.OriginalImage = im.Image
+			im.Image = rn.String()
+			bun.Images[k] = im
+		}
+	}
+
+	return nil
+}
+
+func isOCI(imageType string) bool {
+	return imageType == "" || imageType == "oci"
+}
+
+func isDocker(imageType string) bool {
+	return imageType == "docker"
+}
+
+func keys(images map[string]bundle.Image) []string {
+	keys := []string{}
+	for k := range images {
+		keys = append(keys, k)
+	}
+	return keys
+}
+
+func (r *relocateCmd) setup() (*bundle.Bundle, error) {
+	bundleFile, err := resolveBundleFilePath(r.inputBundle, r.home.String(), r.inputBundleIsFile)
+	if err != nil {
+		return nil, err
+	}
+
+	bun, err := loadBundle(bundleFile)
+	if err != nil {
+		return nil, err
+	}
+
+	if err = bun.Validate(); err != nil {
+		return nil, err
+	}
+
+	return bun, nil
+}
+
+func (r *relocateCmd) writeBundle(bf *bundle.Bundle) (string, error) {
+	data, digest, err := marshalBundle(bf)
+	if err != nil {
+		return "", fmt.Errorf("cannot marshal bundle: %v", err)
+	}
+
+	if r.outputBundleIsFile {
+		if err := ioutil.WriteFile(r.outputBundle, data, 0644); err != nil {
+			return "", fmt.Errorf("cannot write bundle to %s: %v", r.outputBundle, err)
+		}
+		return digest, nil
+	}
+
+	if err := ioutil.WriteFile(filepath.Join(r.home.Bundles(), digest), data, 0644); err != nil {
+		return "", fmt.Errorf("cannot store bundle : %v", err)
+
+	}
+
+	// record the new bundle in repositories.json
+	if err := recordBundleReference(r.home, bf.Name, bf.Version, digest); err != nil {
+		return "", fmt.Errorf("cannot record bundle: %v", err)
+	}
+
+	return digest, nil
+}
+
+type flagsValidator func(cmd *cobra.Command) error
+
+func flagValidRepository(flagName string) flagsValidator {
+	return func(cmd *cobra.Command) error {
+		repositoryValue := cmd.Flag(flagName).Value.String()
+
+		if strings.HasSuffix(repositoryValue, "/") || strings.Contains(repositoryValue, "//") {
+			return fmt.Errorf("invalid repository: %s", repositoryValue)
```

Comment:
As you have here a clear error case, maybe you can improve the error message and tell what is the exact issue in the repository name?

Reply:
Fixed.